rust/stub/src/uefi_helpers.rs

```diff
@@ -1,12 +1,180 @@
+       const XBOOTLDR = 1 << 5;
+       const RandomSeed = 1 << 6;
+       const LoadDriver = 1 << 7;
+       const SortKey = 1 << 8;
+       const SavedEntry = 1 << 9;
+       const DeviceTree = 1 << 10;
+    }
+}
+
+/// Get the SystemdLoaderFeatures if they do exist.
+pub fn get_loader_features(runtime_services: &RuntimeServices) -> Result<SystemdLoaderFeatures> {
+    if let Ok(size) = runtime_services.get_variable_size(cstr16!("LoaderFeatures"), &SD_LOADER) {
+        let mut buffer = vec![0; size].into_boxed_slice();
+        runtime_services.get_variable(
+            cstr16!("LoaderFeatures"),
+            &SD_LOADER,
+            &mut buffer)?;
+
+        return SystemdLoaderFeatures::from_bits(
+            u64::from_le_bytes(
+                (*buffer).try_into()
+                .map_err(|_err| uefi::Status::BAD_BUFFER_SIZE)?
+            ))
+            .ok_or_else(|| uefi::Status::INCOMPATIBLE_VERSION.into());
+    }
+
+    Ok(Default::default())
+}
+
+bitflags! {
+    #[repr(transparent)]
+    #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+    pub struct SystemdStubFeatures: u64 {
+       /// Is `LoaderDevicePartUUID` loaded in UEFI variables?
+       const ReportBootPartition = 1 << 0;
+       /// Are credentials picked up from the boot partition?
+       const PickUpCredentials = 1 << 1;
+       /// Are system extensions picked up from the boot partition?
+       const PickUpSysExts = 1 << 2;
+       /// Are we able to measure kernel image, parameters and sysexts?
+       const ThreePcrs = 1 << 3;
+       /// Can we pass a random seed to the kernel?
+       const RandomSeed = 1 << 4;
+    }
+}
+
+/// Ensures that an UEFI variable is set or set it with a fallback value
+/// computed in a lazy way.
+pub fn ensure_efi_variable<F>(runtime_services: &RuntimeServices,
+    name: &CStr16,
+    vendor: &VariableVendor,
+    attributes: VariableAttributes,
+    get_fallback_value: F) -> uefi::Result
+    where F: FnOnce() -> uefi::Result<Vec<u8>>
+{
+    // If we get a variable size, a variable already exist.
+    if runtime_services.get_variable_size(name, vendor).is_err() {
+        runtime_services.set_variable(
+            name,
+            vendor,
+            attributes,
+            get_fallback_value()?.as_slice()
+        )?;
+    }
+
+    uefi::Status::SUCCESS.into()
+}
+
+/// Exports systemd-stub style EFI variables
+pub fn export_efi_variables(system_table: &SystemTable<Boot>) -> Result<()> {
+    let boot_services = system_table.boot_services();
+    let runtime_services = system_table.runtime_services();
+
+    let stub_features: SystemdStubFeatures =
+        SystemdStubFeatures::ReportBootPartition;
+
+    let loaded_image =
+        boot_services.open_protocol_exclusive::<LoadedImage>(boot_services.image_handle())?;
+
+    let default_attributes = VariableAttributes::BOOTSERVICE_ACCESS | VariableAttributes::RUNTIME_ACCESS;
+
+    // LoaderDevicePartUUID
+    let _ = ensure_efi_variable(runtime_services,
```

Comment:
```suggestion
    ensure_efi_variable(runtime_services,
```

Reply:
Do you mean that error while exporting this variable is fatal?

Reply:
This will generate a warning?

Reply:
Do you want me to disable `unused_must_use`?

Reply:
Ah because it returns a Result. My bad. You could use `.ok()` and then you dont need to assign the resulting `Option` to anything. But I don't have a strong preference in that case.

Reply:
Good point, will do the ok